sw/device/silicon_creator/lib/cert/asn1.c

```diff
@@ -0,0 +1,172 @@
+    // Sanity check: asn1_start_tag should have outputted three bytes.
+    if (tag->len_size != 3)
+      return kErrorAsn1Internal;
+    // Fixup length.
+    tag->state->buffer[tag->len_offset + 0] = 0x82;
+    tag->state->buffer[tag->len_offset + 1] = (uint8_t)(length >> 8);
+    tag->state->buffer[tag->len_offset + 2] = (uint8_t)(length & 0xff);
+  } else {
+    // We only need one byte to hold the length but we allocated three:
+    // we need to shift all the tag data by two bytes. A forward copy
+    // works for that purpose.
+    // TODO: use memmove when we have an implementation.
+    for (size_t i = 0; i < length; i++) {
+      tag->state->buffer[tag->len_offset + 1 + i] =
+          tag->state->buffer[tag->len_offset + tag->len_size + i];
+    }
+    // Fixup length.
+    tag->state->buffer[tag->len_offset] = (uint8_t)length;
+    // Fixed up state offset.
+    tag->state->offset -= tag->len_size - 1;
+  }
+  // Hardening: clear out the tag structure to prevent accident reuse.
+  tag->state = NULL;
+  tag->len_offset = 0;
+  tag->len_size = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_push_bool(asn1_state_t *state, bool value) {
+  // Avoid using the tag functions since everything is constant.
+  RETURN_IF_ERROR(asn1_push_byte(state, ASN1_TAG_BOOLEAN));
+  RETURN_IF_ERROR(asn1_push_byte(state, 1));
+  return asn1_push_byte(state, value ? 0xff : 0);
+}
+
+rom_error_t asn1_push_int32(asn1_state_t *state, uint8_t tag, int32_t value) {
+  uint32_t u_value = (uint32_t)value;
+  uint8_t bigint[4] = {
+      u_value >> 24,
+      (u_value >> 16) & 0xff,
+      (u_value >> 8) & 0xff,
+      u_value & 0xff,
+  };
+  return asn1_push_integer(state, tag, true, bigint, sizeof(bigint));
+}
+
+rom_error_t asn1_push_uint32(asn1_state_t *state, uint8_t tag, uint32_t value) {
+  uint8_t bigint[4] = {
+      value >> 24,
+      (value >> 16) & 0xff,
+      (value >> 8) & 0xff,
+      value & 0xff,
+  };
+  return asn1_push_integer(state, tag, false, bigint, sizeof(bigint));
+}
+
+rom_error_t asn1_push_integer(asn1_state_t *state, uint8_t tag, bool is_signed,
+                              const uint8_t *bytes_be, size_t size) {
+  if (size == 0)
+    return kErrorAsn1InvalidArgument;
+  asn1_tag_t tag_st;
+  RETURN_IF_ERROR(asn1_start_tag(state, &tag_st, tag, size));
+  // Compute smallest possible encoding: ASN1 forbids that the first 9 bits (ie
+  // first octet) and MSB of the second octet are either all ones or all zeroes.
+  size_t offset = 0;
+  // First get rid of unecessary zeroes.
+  while (size >= 2 && bytes_be[offset] == 0 &&
+         (bytes_be[offset + 1] >> 7) == 0) {
+    offset += 1;
+    size -= 1;
+  }
+
+  if (is_signed) {
+    // For signed numbers, get rid of unecessary ones.
+    while (size >= 2 && bytes_be[offset] == 0xff &&
+           (bytes_be[offset + 1] >> 7) == 1) {
+      offset += 1;
+      size -= 1;
+    }
+  } else {
+    // For unsigned numbers, add a 0x00 padding if the first octet has MSB set.
```

Comment:
Perhaps add a bit more explanation here:
```suggestion
    // Integers in ASN.1 are always signed and represented in two's complement.
    // So for unsigned numbers that has MSB set, add a 0x00 padding.
```